File: src/waldiez/models/Agent/Reasoning/ReasoningAgent.ts
/**
 * SPDX-License-Identifier: Apache-2.0
 * Copyright 2024 - 2025 Waldiez & contributors
 */
import { WaldiezAgent, WaldiezNodeAgentType } from "@waldiez/models/Agent/Common";
import { WaldiezAgentReasoningData } from "@waldiez/models/Agent/Reasoning/ReasoningAgentData";

/**
 * WaldiezAgentReasoning
 * @param id - The id of the reasoning agent
 * @param type - The type of the node in a graph (agent)
 * @param agentType - The type of the node in a graph (reasoning)
 * @param name - The name of the agent
 * @param description - The description of the agent
 * @param tags - The tags of the agent
 * @param requirements - The requirements of the agent
 * @param createdAt - The creation date of the agent
 * @param updatedAt - The update date of the agent
 * @param data - The data of the agent. See {@link WaldiezAgentReasoningData}
 * @param rest - Any other data
 */
export class WaldiezAgentReasoning extends WaldiezAgent {
    data: WaldiezAgentReasoningData;
    agentType: WaldiezNodeAgentType = "reasoning";

    constructor(props: {
        id: string;
        agentType: WaldiezNodeAgentType;
        name: string;
        description: string;
        tags: string[];
        requirements: string[];
        createdAt: string;
        updatedAt: string;
        data: WaldiezAgentReasoningData;
        rest?: { [key: string]: unknown };
    }) {
        super(props);
        this.data = props.data;
        this.agentType = "reasoning";
        this.rest = props.rest || {};
    }
}

File: waldiez/models/agents/reasoning/reasoning_agent.py
# SPDX-License-Identifier: Apache-2.0.
# Copyright (c) 2024 - 2026 Waldiez and contributors.

# pyright: reportArgumentType=false,reportIncompatibleVariableOverride=false

"""Reasoning agent model."""

from typing import Any

from pydantic import Field
from typing_extensions import Annotated, Literal

from ..agent import WaldiezAgent
from .reasoning_agent_data import WaldiezReasoningAgentData


class WaldiezReasoningAgent(WaldiezAgent):
    """Reasoning agent model."""

    agent_type: Annotated[
        Literal["reasoning"],
        Field(
            "reasoning",
            title="Agent type",
            description="The agent type in a graph: 'reasoning'",
            alias="agentType",
        ),
    ]
    data: Annotated[
        WaldiezReasoningAgentData,
        Field(
            title="Data",
            description="The reasoning agent's data",
            default_factory=WaldiezReasoningAgentData,
        ),
    ]

    def get_reasoning_config(self) -> dict[str, Any]:
        """Get the reasoning configuration based on its method.

        Returns
        -------
        dict
            The reasoning configuration.
        """
        return self.data.get_reasoning_config()
